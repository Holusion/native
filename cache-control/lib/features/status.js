

export const INITIAL_LOAD = "INITIAL_LOAD";
export const SET_NETINFO = 'SET_NETINFO';
export const SET_SIGNEDIN = "SET_SIGNEDIN";
export const SET_SYNCHRONIZED = "SET_SYNCHRONIZED";

export default function status(state = {
  initial_load: false,
  connected: false,
  signedIn: false,
  synchronized: false,
}, { type, error, value=true}) {
  switch (type) {
    case INITIAL_LOAD:
      //Initial load is always set to true even if it had an error
      return { ...state, initial_load: !!value };
    case SET_NETINFO:
      return {...state, connected: !!value};
    case SET_SIGNEDIN:
      return {...state, signedIn: error? false : !!value};
    case SET_SYNCHRONIZED:
      return {...state, synchronized: error? false : !!value};
    default:
      return state;
  }
}

export const isLoaded = (state) => state.status.initial_load;
export const isConnected = state => state.status.connected;
export const isSignedIn = (state) => state.status.signedIn;
export const isSynchronized = (state)=> state.status.synchronized;

/**
 * 
 * @param {('offline'|'online')} connected
 */
export const setNetInfo = (connected)=>{
  return {type: SET_NETINFO, value: connected}
}

export const setSignedIn = (value)=>{
  if(value instanceof Error) {
    return {type: SET_SIGNEDIN, error: value};
  }
  return {type: SET_SIGNEDIN, value};
}

/**
 * 
 * @param {boolean|Error} value 
 * @return {import("redux").Action}
 */
export const setSynchronized = (value) =>{
  if(value instanceof Error) {
    return {type: SET_SYNCHRONIZED, error: value};
  }
  return {type: SET_SYNCHRONIZED, value};
}