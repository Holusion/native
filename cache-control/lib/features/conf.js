'use strict';
import { INITIAL_LOAD } from "./status";

export const actions = {
    SET_CONF:  "SET_CONF",
    SET_SLIDES_CONTROL: "SET_SLIDES_CONTROL",
    SET_PLAY_CONTROL: "SET_PLAY_CONTROL",
    SET_DEFAULT_TARGET: "SET_DEFAULT_TARGET",
    SET_PURGE: "SET_PURGE",
    SET_PASSCODE: "SET_PASSCODE",
    SET_PROJECTNAME: "SET_PROJECTNAME",
    SET_TIMEOUT: "SET_TIMEOUT",
    SET_WATCH: "SET_WATCH",
    SET_AUTOCLEAN: "SET_AUTOCLEAN",
}
export const action_strings = Object.keys(actions).reduce((res, k)=>([...res, actions[k]]),[]);
/**
 * @typedef {object} AppConf
 * @property {string} [projectName] 
 * @property {boolean} configurableProjectName: true, // This is read-only from initialization
 * @property {string|null} default_target: null,
 * @property {boolean} purge_products: false,
 * @property {"default"|"buttons"|"swipe"|"none"} slides_control: "default",
 * @property {"button"|"rotate"|"none"} play_control: "none",
 * @property {boolean} watch: true,
 * @property {number} timeout: 0,
 * @property {boolean} autoClean: false,
 */

/**
 * Local app configuration.
 * Not the same as state.data.config, 
 * which is the global "application" configuration from database (ie. the `applications/${appId}` doc)
 * @param {*} state 
 * @param {*} action 
 */
export default function conf(state = {
    projectName: undefined, 
    configurableProjectName: true, // This is read-only from initialization
    default_target: null,
    purge_products: false,
    slides_control: "default",
    play_control: "none",
    watch: true,
    timeout: 0,
    autoClean: false,
}, action) {
    if(action.error){
        return state;
    }
    switch(action.type) {
      case INITIAL_LOAD:
      case actions.SET_CONF:
        return Object.assign({}, state, action.conf);
      case actions.SET_DEFAULT_TARGET:
        return Object.assign({}, state, {default_target: action.target});
      case actions.SET_PURGE:
        return Object.assign({}, state, {purge_products: action.purge});
      case actions.SET_SLIDES_CONTROL:
        return Object.assign({}, state, {slides_control: action.control});
      case actions.SET_PLAY_CONTROL:
        return Object.assign({}, state, {play_control: action.control});
      case actions.SET_PROJECTNAME:
        if(!state.configurableProjectName){
            console.warn("Trying to modify read-only project name");
            return state;
        }
        return Object.assign({}, state, {projectName: action.projectName});
      case actions.SET_PASSCODE:
        return Object.assign({}, state, {passcode: action.passcode});
      case actions.SET_WATCH:
        return Object.assign({}, state, {watch: action.watch});
      case actions.SET_TIMEOUT:
        return Object.assign({}, state, {timeout: action.timeout});
      case actions.SET_AUTOCLEAN:
        return Object.assign({}, state, {autoClean: action.autoClean})
      default:
        return state;
    }
}

/**
 * 
 * @param {AppConf["slides_control"]} control - one of default, buttons, none
 */
export const setSlidesControl = (control) =>{
    return {type : actions.SET_SLIDES_CONTROL, control}
}
/**
 * 
 * @param {AppConf["play_control"]} control - one of button, rotate, none
 */
export const setPlayControl = (control) =>{
    return {type : actions.SET_PLAY_CONTROL, control}
}

export const setDefaultTarget = (name)=>{
    return {type: actions.SET_DEFAULT_TARGET, target: name};
}
export const setPurge = (purge)=>{
    return {type: actions.SET_PURGE, purge: purge};
}

export const getProjectName = (state)=> state.conf.projectName; 
export const setProjectName = (projectName) => {
    return {type: actions.SET_PROJECTNAME, projectName};
}

export const setPasscode = (passcode) => {
    return {type: actions.SET_PASSCODE, passcode};
}

export const getWatch = (state) => state.conf.watch;
export const setWatch = (watch)=>{
  return {type: actions.SET_WATCH, watch: !!watch};
}

export const setTimeout = (timeout)=>{
  return {type: actions.SET_TIMEOUT, timeout};
}
/**
 * Automatically clean cache's unused files
 * @param {boolean} autoClean 
 * @returns 
 */
export const setAutoClean = (autoClean)=>{
  return {type: actions.SET_AUTOCLEAN, autoClean};
}

export const getAutoClean = (state)=> state.conf.autoClean;

export const getConf = (state)=>{
    return state.conf;
}

/**
 * Completely overwrite conf. Only for initial load.
 */
export const setConf = ({error, ...conf})=>{
    if(error){
        return {type: actions.SET_CONF, error};
    }
    return {type: actions.SET_CONF, conf};
}
