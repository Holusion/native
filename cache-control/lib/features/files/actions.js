import { INITIAL_LOAD } from "../status";

export const SET_FILES = "SET_FILES";
export const SET_DEPENDENCIES = "SET_DEPENDENCIES";
export const SET_CACHED_FILE = "SET_CACHED_FILE";
export const UNSET_CACHED_FILE = "UNSET_CACHED_FILE";


export const cacheFile = "cache_v1";

export const getFiles = (state)=>state.files.list;
export const setFiles = ({error, ...files}) =>{
  if(error){
    return {type: INITIAL_LOAD, error}
  }
  return {type: INITIAL_LOAD, files}
};


export const getDependencies = (state)=> Object.keys(state.files.list);
/**
 * 
 * @param {string="items","config"} name 
 * @param {object} list - an dict of dest->FileRef 
 */
export const setDependencies = (name, list)=>({type: SET_DEPENDENCIES, name, list});

/**
 * 
 * @param {*} state 
 * @param {string} file 
 */
export const getHash = (state, file) => getCache(state)[
  ((file.indexOf("file://") === 0) ? file.slice("file://".length) : file)
];

/**
 * 
 * @param {string} filename path to the file (absolute path is generally used)
 * @param {string} hash the file's hash. base64-encoded md5 is generally assumed
 * @returns 
 */
export const setHash = (filename, hash) => ({type:SET_CACHED_FILE, file: filename, hash});
/**
 * Remove a file from the index
 * @param {string} filename path to the file. One of the keys in `state.files.cache`
 * @returns 
 */
export const unsetHash = (filename)=>({type:UNSET_CACHED_FILE, file: filename});

/**
 * Get the whole cache state (internal use)
 * @param {*} state 
 */
export const getCache = (state)=> state.files.cache;

export const isCached = (state, name, hash)=> getCache(state)[name] === (hash || true);

/**
 * All file dependencies in cache and fresh
 * @param {import("redux").Store} state 
 * @returns {string[]} a list of file paths
 */
export const getCachedFiles = (state) => Object.keys(state.files.list).filter(f=> {
  return state.files.cache[f] && state.files.cache[f] === state.files.list[f].hash
});

/**
 * all file dependencies not in cache or stale
 * @param {import("redux").Store} state 
 * @returns {string[]}
 */
export const getUncachedFiles = (state) => Object.keys(state.files.list).filter(f=> {
  return !state.files.cache[f] || state.files.cache[f] !== state.files.list[f].hash
});


/**
 * filter out files that are not required for initial render
 * ie. videos
 * @returns {boolean}
 */
export const isRequired = (f)=> !/^video\//.test(f.contentType)

export const getRequiredFiles = (state) => getUncachedFiles(state).filter(f => isRequired(state.files.list[f]));
export const getOtherFiles = (state) => getUncachedFiles(state).filter(f=> !isRequired(state.files.list[f]));

export const getTotalSize = (state) => Object.keys(state.files.list).reduce((size, file) => (size + (state.files.list[file].size)), 0);
export const getUncachedSize = (state) => getUncachedFiles(state).reduce((size, file) => (size + (state.files.list[file].size)), 0);
export const getRequiredSize = (state)=> getRequiredFiles(state).reduce((size, file) => (size + (state.files.list[file].size)), 0);
export const getOtherSize = (state)=> getOtherFiles(state).reduce((size, file)       => (size + (state.files.list[file].size)), 0);

export const isBlocked = (state) => getRequiredFiles(state).length !== 0;