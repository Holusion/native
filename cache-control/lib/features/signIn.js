
import {firebase, auth} from "firebase";

import { getUniqueId, getApplicationName, getDeviceName } from "deviceInfo";

import { put, delay, call, select } from 'redux-saga/effects'

import {setSignedIn, SET_SIGNEDIN} from "./status";

import {warn} from "./logs";
import { getProjectName } from "./conf";

export const DO_SIGNIN = "DO_SIGNIN";

export async function doSignIn(projectName){
  const hostname = await getDeviceName();
  const uuid = await getUniqueId(); //Unique and reasonably persistent
  //Ridiculously low entropy but we just need to avoid possible hostname collision
  const shortId = uuid.slice(0, 8);
  try{
    const func = firebase.app().functions("europe-west1").httpsCallable("https_authDeviceCall")
    let { data: token } = await func({ 
      uuid, //Used as a "password"
      applications: [projectName], 
      meta: { publicName: `${getApplicationName()}.${hostname}#${shortId}`} 
    });
    await auth().signInWithCustomToken(token);
    return null;
  }catch(e){
    return e;
  }
}


export function* signIn(){
  let d = 512;
  const projectName = yield select(getProjectName);
  if(!projectName) return;
  yield put(setSignedIn(false));
  while(true){
    let error = yield call(doSignIn, projectName);
    if(!error) break;
    if(/internet.*offline/i.test(error.message)){
      //httpsCallable error from firebase iOS SDK has no code for "offline"
      //It yield an useless error with code unknown
      yield put(warn(SET_SIGNEDIN, "Impossible de se connecter", "la tablette n'est probablement pas reliée à internet"))
    }else{
      yield put(setSignedIn(error));
    }
    yield delay(d);
    /* istanbul ignore next */
    if(d <= 32768){
      d = d*2;
    }else{
      return;
    }
  }
  yield put(setSignedIn(projectName));
}

export function trySignIn(){
  return {type: DO_SIGNIN};
}