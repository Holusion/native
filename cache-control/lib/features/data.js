'use strict';
import { eventChannel } from '@redux-saga/core';
import { select, take, call, put } from "redux-saga/effects";
import {createSelector} from "reselect";


import {WatchChanges} from "../WatchChanges";
import { getProjectName, getWatch } from './conf';
import { handleSetData } from './files/handleSetData';
import { INITIAL_LOAD, SET_SIGNEDIN } from './status';


export const GET_DATA = "GET_DATA";
export const SET_DATA = "SET_DATA";

export default function data(state = {
  items: {}, //set in bulk by database applications/:appId/pages
  config: {}, //set in bulk by applications/appId
}, { type, data, error }) {
  if (error) {
    return state;
  }
  switch (type) {
    case INITIAL_LOAD:
    case SET_DATA:
      return { ...state, ...data };
    default:
      return state;
  }
}


export const getData = (state)=> state.data;
/**
 * set internal data (items or config) only one of action.items or action.config should be set.
 * @param {object} action 
 * @param {Error} [action.error] an error to be logged (data state will not be changed)
 * @param {object} [action.items] new items
 * @param {object} [action.config] project configuration
 * @param {object|Map<string, import('../WatchChanges').FileRef>} [action.files] associated files dependencies
 * 
 * @returns {object}
 */
export const setData = ({ files, error, ...data }) => {
  if(error){
    return { type: SET_DATA, error };
  }
  return { type: SET_DATA, files, data };
}

export const getConfig = (state)=> getData(state).config;
export const getItems = (state) => state.data.items;


export const getItemsIds = createSelector(
    [getItems],
    (items) => Object.keys(items)
)

export const getItemsArray = createSelector(
    [getItemsIds, getItems],
    (ids,items) => ids.map(id => Object.assign({id}, items[id]) ),
)

/**
 * simple wrapper around WatchChanges to be called by redux-saga
 * @param {string} projectName 
 * @see WatchChanges
 */
export function createWatcher(projectName){
  return new WatchChanges({projectName});
}

export function createWatchChannel(wc, watch=true) {
  return eventChannel( (emit) => {
    wc.on("error", (err)=> emit({error: err}) );
    wc.on("dispatch", (data)=> emit(data) );
    
    if(watch) {
      wc.watch();
    }else {
      wc.getOnce();
    }
    return ()=> {
      wc.close();
      wc.removeAllListeners();
    };
  });
}

export function* watchChanges(action) {
  if(action.error) return;
  const projectName = action.type === SET_SIGNEDIN? action.value : yield select(getProjectName);
  const doWatch = yield select(getWatch);
  if(!projectName) return;
  const wc = yield call(createWatcher, projectName);
  const watchChannel = yield call(createWatchChannel, wc, doWatch);
  try{
    while(true){
      const payload = yield take(watchChannel);
      const action = setData(payload);
      //Payload can also be an error, in which case we may need to close the loop?
      //Set file dependencies BEFORE we dispatch SET_DATA
      yield* handleSetData(action);
      yield put(action);
    }
  }finally{
    watchChannel.close();
  }
}