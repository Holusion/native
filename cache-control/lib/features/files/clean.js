import { call, select, put, takeLatest } from 'redux-saga/effects'

import fs from "filesystem";
import { basename } from "filepaths";

import  { info, warn } from "../logs";

import { getCache, getFiles, unsetHash } from "./actions";

export const CLEAN_CACHE = "CLEAN_CACHE";

/**
 * Walk files in state.files.cache to remove unused ones
 * Integrated with `takeLatest(CLEAN_CACHE, cleanCache);`
 */
export function* cleanCache(){
  const cache = yield select(getCache);
  const dependencies = yield select(getFiles);
  const unused_files = Object.keys(cache).filter(name=> typeof dependencies[name] === "undefined");
  let errors = 0;
  for(let file of unused_files){
    try{
      yield call(fs.unlink, file);
      yield put(unsetHash(file));
      yield put(info(CLEAN_CACHE, `${basename(file)} supprimé du cache (inutilisé)`));
    }catch(e){
      if(e.code ==="ENOENT"){
        yield put(unsetHash(file));
      }else {
        errors++;
      }
      yield put(warn(CLEAN_CACHE, `Erreur lors de la suppression de ${basename(file)}: ${e.message}`));
    }
    if(errors ==0){
      yield put(info(CLEAN_CACHE, `cache nettoyé (${unused_files.length} fichiers supprimés)`));
    }
  }
}