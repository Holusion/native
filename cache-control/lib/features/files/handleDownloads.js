
import { put, takeLatest, take, takeEvery, cancelled, select, call, fork } from "redux-saga/effects";

import writeToFile from "writeToFile";
import { info } from "../logs";

import {SET_DEPENDENCIES, getRequiredFiles, getOtherFiles, setHash, SET_CACHED_FILE, getFiles, isCached} from "./actions";


//Used only internally. Otherwise it would be defined in ./actions.js
export const REQUEST_DOWNLOAD = "REQUEST_DOWNLOAD";
export const DO_DOWNLOAD = "DO_DOWNLOAD";
/**
 * Meant to be used with takeLatest(SET_DEPENDENCIES, handleDownload)
 * */
export function* schedule_downloads(){
  const required = yield select(getRequiredFiles);
  const other = yield select(getOtherFiles);
  const list = yield select(getFiles);
  for(let files of [required, other]){
    for(let dest of files){
      const file = list[dest];
      yield put({type: REQUEST_DOWNLOAD, dest: dest, src:file.src, hash: file.hash});
      while(true){
        let cached = yield take(SET_CACHED_FILE);
        if(cached.file === dest && cached.hash === file.hash) break;
      }
    }
  }
}

export function* do_download(){
  while(true){
    const {src, dest, hash} = yield take(REQUEST_DOWNLOAD);
    //Check if file was downloaded between the time request was issued and now
    if( !( yield select(isCached, dest, hash) )){
      yield put(info(DO_DOWNLOAD, `Téléchargement de ${dest.split("/").slice(-1)[0]}`));
      try{
        yield call(writeToFile, src, dest);
      }catch(e){
        yield put({type:DO_DOWNLOAD, error: e});
        return;
      }
    }
    yield put(setHash(dest, hash));
  }
}

export function* handleDownloads(){
  yield fork(do_download);
  yield call(schedule_downloads);
  yield takeLatest(SET_DEPENDENCIES, schedule_downloads);
}