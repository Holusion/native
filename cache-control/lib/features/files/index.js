'use strict';


import {SET_DEPENDENCIES, SET_CACHED_FILE, UNSET_CACHED_FILE} from "./actions";
import {INITIAL_LOAD} from "../status";

export * from "./actions";
export {CLEAN_CACHE, cleanCache} from "./clean";
export {handleDownloads} from "./handleDownloads";
export {handleSetData} from "./handleSetData";


export default function files(state = {
  list:{
  },
  cache: {},
  itemFiles: [],
  configFiles: [],
}, action ) {
  if(action.error) return state;
  switch(action.type){
    case INITIAL_LOAD:
      return {...state, ...action.files};
    case SET_DEPENDENCIES:
      let list = action.list
      let itemFiles = action.name === "items"? Object.keys(list) : state.itemFiles;
      let configFiles = action.name === "config"? Object.keys(list) : state.configFiles;
      let dependencies = [].concat(
        itemFiles,
        configFiles,
      )

      let raw_list = {...state.list, ...list};
      let dep_list = {};
      for (let f in raw_list){
        if(dependencies.indexOf(f) !== -1){
          dep_list[f] = raw_list[f];
        }
      }

      return {
        cache: state.cache,
        list: dep_list,
        itemFiles,
        configFiles,
      }
    case SET_CACHED_FILE:
      return {...state, cache:{...state.cache,[action.file]: action.hash}}
    case UNSET_CACHED_FILE:
      const {[action.file]:removed_file, ...files} = state.cache
      return {...state, cache: files}

    default:
      return state;
  }
}
