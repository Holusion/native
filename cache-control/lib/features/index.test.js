import {runSaga} from "redux-saga";
import { takeLatest, throttle } from "redux-saga/effects";
import {testSaga, expectSaga} from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';

jest.mock("../path");
import fsMock from "filesystem";
import {app as appMock, auth as autoMock} from "firebase";

import {loadLocalSaga, reducers, rootSaga, sagaStore, saveCache, dataFile, getPersistentState, info, getUncachedFiles, CLEAN_CACHE} from ".";
import {SET_DATA, watchChanges, createWatcher} from "./data";
import {setConf, actions, action_strings as conf_actions_names, getProjectName, getWatch, setProjectName, getAutoClean} from "./conf";

import { signIn, doSignIn } from "./signIn";
import { storagePath } from "../path";

import {makeFileRef} from "./files/_mock_fileRef";
import { INITIAL_LOAD, isLoaded, setSignedIn } from "./status";
import { saveFile } from "../readWrite";

afterEach(()=>{
  fsMock._reset();
})

describe("rootSaga", ()=>{

  test("will auth initially if projectName is set", ()=>{
    let s = reducers(undefined, setConf({projectName:"foo"}))
    return expectSaga(rootSaga)
    .withReducer(reducers, s)
    .provide([
      [matchers.call.fn(loadLocalSaga), undefined],
    ])
    .call(doSignIn, "foo")
    .silentRun()
  })

  test("will reset change listeners on re-auth", ()=>{
    return expectSaga(rootSaga)
    .withReducer(reducers, )
    .provide([
      [matchers.call.fn(loadLocalSaga), undefined],
    ])
    .dispatch(setSignedIn("foo"))
    .call(createWatcher, "foo")
    .dispatch(setSignedIn("bar"))
    .call(createWatcher, "bar")
    .silentRun()
  })

  test("will save conf changes to disk", ()=>{
    return expectSaga(rootSaga)
    .withReducer(reducers)
    .provide({
      call(effect, next){
        if(effect.fn === loadLocalSaga) return undefined;
        next();
      },
      fork(effect, next){
        if(effect.fn.name == "debounceHelper") return takeLatest(...effect.args.slice(1));
        next();
      }
    })
    .dispatch({type: actions.SET_PROJECTNAME, projectName: "foo"})
    .delay(600)
    .fork.fn(saveCache)
    .silentRun()
  })
})

describe("saveCache()", ()=>{
  test("handle saveFile error", ()=>{
    let s = {
      files: "files_foo",
      data: "data_foo",
      conf: "conf_foo"
    }
    let e = new Error("Booh");
    testSaga(saveCache).next()
    .select(getPersistentState).next(s)
    .call(JSON.stringify, s).next("stringified_state")
    .call(saveFile, dataFile, "stringified_state").throw(e)
    .put({type:"SAVE_CACHE", error: e}).next()
    .isDone();
  });

  test("can save data file", ()=>{

    testSaga(saveCache, {type: SET_DATA}).next()
    .select(getPersistentState).next({})
    .call(JSON.stringify, {}).next("stringified_state")
    .call(saveFile, dataFile, "stringified_state").next()
    .put(info("SAVE_CACHE", `Données locales sauvegardées`, `Déclenché par ${SET_DATA}`)).next()
    .select(getAutoClean).next(false)
    .isDone();
  });
  describe("cache cleanup", ()=>{
    let saga;
    beforeEach(()=>{
      saga = testSaga(saveCache, {type: SET_DATA}).next()
      .select(getPersistentState).next({})
      .call(JSON.stringify, {}).next("stringified_state")
      .call(saveFile, dataFile, "stringified_state").next()
      .put(info("SAVE_CACHE", `Données locales sauvegardées`, `Déclenché par ${SET_DATA}`)).next()
      .select(getAutoClean).next(true)
      .delay(5000).next();
    });

    test("won't trigger cleanup if some files are not cached", ()=>{
      saga.select(getUncachedFiles).next(["/tmp/uncached.txt"])
      .isDone();
    });
    test("will trigger cleanup", ()=>{
      saga.select(getUncachedFiles).next([])
      .put({type: CLEAN_CACHE}).next()
      .isDone();
    });
  })
})

describe("loadLocalSaga", function(){
  test("dataFile is properly set", ()=>{
    expect(dataFile).toMatch(/data_v\d+\.json/i);
  })
  test("Load local files", async ()=>{
    fsMock._set(`${storagePath()}/${dataFile}`, JSON.stringify({
      data: {config: {foo: "bar"}},
      conf: {projectName: "foofoo"},
      files: {
        list:{},
        cache: {"/tmp/list.png":"xxxxxx"}
      }
    }));

    const result = await expectSaga(loadLocalSaga)
    .put({
      type: INITIAL_LOAD, 
      data: {config: {foo: "bar"}},
      conf: {projectName: "foofoo"},
      files: { list:{}, cache: {"/tmp/list.png":"xxxxxx"}}
    })
    .run()
    expect(result.storeState).toMatchSnapshot();
  });
  
  test("Handle local files absence", async ()=>{
    let e = new Error(`no such file or directory`);
    e.code = "ENOENT";
    fsMock.readFile.mockImplementation(()=>Promise.reject(e));
    await expectSaga(loadLocalSaga)
    .put({ type: 'LOG_INFO',
      name: 'INITIAL_LOAD',
      message: 'data_v1.json n\'existait pas',
      context: 'Soit c\'est une nouvelle installation soit le fichier a été perdu' 
    })
    .run();
  });
  test("Handle EACCESS error", async ()=>{
    let e = new Error(`permission denied`);
    e.code = "EACCESS";
    fsMock.readFile.mockImplementation(()=>Promise.reject(e));
    await expectSaga(loadLocalSaga)
    .put({type: INITIAL_LOAD, error: new Error(`Could not load ${dataFile} : permission denied`)})
    .run()
  });
})

describe("sagaStore()", ()=>{
  let func = jest.fn(), signInWithCustomToken = jest.fn();
  beforeEach(()=>{
    signInWithCustomToken.mockReset();
    signInWithCustomToken.mockImplementation(()=>Promise.resolve());
    autoMock.mockImplementation(()=>({
      signInWithCustomToken,
    }))
    appMock.mockImplementation(()=>({
      functions: jest.fn(()=>({
        httpsCallable: jest.fn(()=>func)
      }))
    }));
  });
  test("creates a connected store on initial load", ()=>{
    const [store] = sagaStore();
    return expect(Promise.race([
      new Promise((_, reject)=> setTimeout(()=>reject(new Error(`store does not match required shape : ${JSON.stringify(store.getState())}`), 100))),
      new Promise(resolve => {
        const unsubscribe = store.subscribe(()=>{
          let state = store.getState()
          if(state.status.initial_load) {
            unsubscribe();
            resolve(state);
          };
        })

      }),
    ])).resolves.toEqual(expect.objectContaining({
      //Maybe do assertions on initial store state
    }));
  });

  test("can set a pre-defined default project name", ()=>{
    const [store, task] = sagaStore({projectName: "holodemo"});
    expect(getProjectName(store.getState())).toEqual("holodemo");
    task.cancel();
  });
  
  test("can change autoClean value", ()=>{
    const [store, task] = sagaStore({autoClean: true});
    expect(getAutoClean(store.getState())).toEqual(true);
    task.cancel();
  });

  describe("after INITIAL_LOAD", ()=>{
    //Some smoke test to check if store properly initializes
    let store, task;
    beforeEach(()=>{
      jest.useFakeTimers();
      ([store, task] = sagaStore());
      return new Promise((resolve)=>{
        const unsubscribe = store.subscribe(()=>{
          const state = store.getState();
          if(isLoaded(state)) {
            unsubscribe();
            resolve();
          }
        });
      });
    });

    afterEach(()=>{
      task.cancel();
    })

    async function doDebounce(){
      jest.advanceTimersByTime(500);
      await Promise.resolve();
    }
    test("save data changes to disk", async ()=>{
      store.dispatch({type:SET_DATA, data: {items: {foo: {title: "bar"}}}});
      await doDebounce();
      expect(fsMock.contents).toMatchSnapshot();
    });

    test("save conf changes to disk", async ()=>{
      store.dispatch(setConf({projectName: "foo", autoClean: true}));
      await doDebounce();
      expect(store.getState()).toHaveProperty("conf", expect.objectContaining({
        projectName:"foo",
        autoClean: true,
        slides_control: "default",
      }))
      expect(fsMock.contents).toMatchSnapshot();
    });

    test('save required files list to disk', async ()=>{
      store.dispatch({
        type:SET_DATA, 
        data: {items: {foo: {image: "/tmp/foo.png"}}}, 
        files: {
          "/tmp/foo.png": makeFileRef("foo.png")
        }
      });
      await doDebounce();
      expect(fsMock.contents[`${storagePath()}/${dataFile}`]).toBeTruthy();
      expect(fsMock.contents[`${storagePath()}/${dataFile}`]).toMatchSnapshot();
    });
  })
})