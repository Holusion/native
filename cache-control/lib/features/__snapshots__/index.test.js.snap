// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadLocalSaga Load local files 1`] = `undefined`;

exports[`sagaStore() after INITIAL_LOAD save conf changes to disk 1`] = `
Object {
  "/tmp/storage/data_v1.json": "{\\"files\\":{\\"list\\":{},\\"cache\\":{},\\"itemFiles\\":[],\\"configFiles\\":[]},\\"data\\":{\\"items\\":{},\\"config\\":{}},\\"conf\\":{\\"projectName\\":\\"foo\\",\\"configurableProjectName\\":true,\\"default_target\\":null,\\"purge_products\\":false,\\"slides_control\\":\\"default\\",\\"play_control\\":\\"none\\",\\"watch\\":true,\\"timeout\\":0,\\"autoClean\\":true}}",
}
`;

exports[`sagaStore() after INITIAL_LOAD save data changes to disk 1`] = `
Object {
  "/tmp/storage/data_v1.json": "{\\"files\\":{\\"list\\":{},\\"cache\\":{},\\"itemFiles\\":[],\\"configFiles\\":[]},\\"data\\":{\\"items\\":{\\"foo\\":{\\"title\\":\\"bar\\"}},\\"config\\":{}},\\"conf\\":{\\"configurableProjectName\\":true,\\"default_target\\":null,\\"purge_products\\":false,\\"slides_control\\":\\"default\\",\\"play_control\\":\\"none\\",\\"watch\\":true,\\"timeout\\":0,\\"autoClean\\":false}}",
}
`;

exports[`sagaStore() after INITIAL_LOAD save required files list to disk 1`] = `"{\\"files\\":{\\"list\\":{},\\"cache\\":{},\\"itemFiles\\":[],\\"configFiles\\":[]},\\"data\\":{\\"items\\":{\\"foo\\":{\\"image\\":\\"/tmp/foo.png\\"}},\\"config\\":{}},\\"conf\\":{\\"configurableProjectName\\":true,\\"default_target\\":null,\\"purge_products\\":false,\\"slides_control\\":\\"default\\",\\"play_control\\":\\"none\\",\\"watch\\":true,\\"timeout\\":0,\\"autoClean\\":false}}"`;
